internal/error.go

```diff
@@ -129,6 +129,16 @@ type (
 		details      converter.EncodedValues
 	}
 
+	ApplicationErrorWithRetryDelay struct {
```

Comment:
I wouldn't put the retry on an error class if we're not serializing it as part of ApplicationFailure.
I still think we want to add the field in ApplicationFailureInfo even if other disagree.

If we must go with the alternative approach, you would probably want to export a `SetNextRetryDelayOnFailure` method on the `activity` package or maybe something more general that could set other attributes as well (although, even though other seem to think so, I disagree that we need to support setting other activity attributes **on failure**).

Reply:
+1 to `SetNextRetryDelayOnFailure` method on the activity package

Reply:
@bergundy thank you for the recommendation.

> I wouldn't put the retry on an error class if we're not serializing it as part of ApplicationFailure.
> I still think we want to add the field in ApplicationFailureInfo even if other disagree.

ApplicationErrorWithRetryDelay is not an error class. You cannot restore one from a failure. ApplicationErrorWithRetryDelay is an entity which implements Error interface and Delay interface.
You see what I did, I labeled a tail as a leg and now claim that a dog has 5 legs 😄 
But seriously, please help me better understand the reason you think we should not put it on the error.
Do you mean that activity to SDK protocol should define SDK to the server protocol?

> If we must go with the alternative approach, you would probably want to export a SetNextRetryDelayOnFailure method on the activity package or maybe something more general that could set other attributes as well (although, even though other seem to think so, I disagree that we need to support setting other activity attributes on failure).

I let user return 2 things which are related in one call: ApplicationError and RetryDelay. I agree there is some issue that what the call produces is of type error, but I did not want to change the type of what Executor expects for backward computability. Regarding other attributes, they may be set by the user in the way which SDK team finds appropriate. They do not have to follow this if the attribute does not relate to the failure.

Reply:
> But seriously, please help me better understand the reason you think we should not put it on the error.

IMO the same reason as was suggested in the discussions on this topic that it should not be on the error in the protos - it should not be a product of the error but a settable property anywhere in the activity for when an error occurs.

+1 to SetNextRetryDelayOnFailure method on the activity package

Reply:
As mentioned when we met - I think it can actually go both places and that makes good sense. I think it makes sense to have an API to attach a next delay to any error you're returning, and also the general `SetNextRetryDelayOnFailure` API. Both can be quite useful depending on the structure of your code, and both can end up in the same place in the actual API response.

Reply:
If we do want to attach it to the error I think we should have something more general then forcing users to use `ApplicationErrorWithRetryDelay`. I would suggest we check an interface like `retryDelay()` and if the error implements that interface we use it. If we allow both I am not sure which takes precedent.

Reply:
+1 - in my mind you would use an interface just like you've said, which you could augment any error with something that wraps the original and implements that interface. As for precedence, I imagine the per-error one would take precedence over the general "next delay" call, since the former is more specific/granular than the latter.

Reply:
It was a mistake by the original developers to make `temporal.NewApplicationError` return `error` instead of `*ApplicationError` and thus we can't really attach things easily to it. Luckily our other SDKs don't suffer from this. I am not a fan of having two ways personally, but I don't have strong objection so long as the easier, simpler, clearer way everyone agrees (`activity.SetNextRetryDelayOnFailure`) exists.

I don't suspect there will be many users writing errors conforming to that interface. Looking at the failure converter, using the interface approach, how can I set error type, some details, and next retry delay in my error? Can I see a snippet of code of how that might look?

Reply:
>Looking at the failure converter, using the interface approach

The failure converter is not relevant. If you look at the current `extractActivityRequestsFrom` it is called on the error returned from the activity. I am suggesting that function should check for an interface rather then a specific type. I don't want to give our types any special behaviour that users types can't have.

Reply:
Can you give an example of returning an error from an activity with this new API with error type, some details, and next retry delay set on the application error? That way I can at least visualize what is meant with the interface approach.